test/distributed/org/apache/cassandra/distributed/upgrade/MixedModeAvailabilityTest.java

```diff
@@ -0,0 +1,137 @@
+                         new AvailabiltyTester(ALL, ONE));
+    }
+
+    public void testAvailability(AvailabiltyTester... testers) throws Throwable
+    {
+        new TestCase()
+        .nodes(3)
+        .nodesToUpgrade(1)
+        .upgrade(Versions.Major.v22, Versions.Major.v30)
+        .upgrade(Versions.Major.v3X, Versions.Major.v4)
+        .upgrade(Versions.Major.v30, Versions.Major.v4)
+        .withConfig(config -> config.set("read_request_timeout_in_ms", SECONDS.toMillis(30))
+                                    .set("write_request_timeout_in_ms", SECONDS.toMillis(30)))
+        .setup(c -> c.schemaChange(withKeyspace("CREATE TABLE %s.tbl (k uuid, c int, v int, PRIMARY KEY (k, c))")))
+        .runAfterNodeUpgrade((cluster, n) -> {
+
+            // using an upgraded and a not upgraded coordinator...
+            for (int node = 1; node < cluster.size(); node++)
+            {
+                ICoordinator coordinator = cluster.coordinator(node);
+                int nodeDown = node;
+
+                // using 0 to 2 down nodes...
+                for (int numNodesDown = 0; numNodesDown < cluster.size(); numNodesDown++)
+                {
+                    // disable communications to the down nodes
+                    if (numNodesDown > 0)
+                    {
+                        nodeDown = nextNode(nodeDown, cluster.size());
+                        cluster.filters().verbs(READ_REQ.id, MUTATION_REQ.id).to(nodeDown).drop();
+                    }
+
+                    // run the test cases that are compatible with the number of down nodes
+                    for (AvailabiltyTester tester : testers)
+                    {
+                        tester.test(coordinator, numNodesDown);
+                    }
+                }
+
+                // restore communication to all nodes
+                cluster.filters().reset();
+            }
+        }).run();
+    }
+
+    private static int nextNode(int node, int numNodes)
+    {
+        return node == numNodes ? 1 : node + 1;
+    }
+
+    private static class AvailabiltyTester
+    {
+        private static final String INSERT = withKeyspace("INSERT INTO %s.tbl (k, c, v) VALUES (?, ?, ?)");
+        private static final String SELECT = withKeyspace("SELECT * FROM %s.tbl WHERE k = ?");
+
+        private final ConsistencyLevel writeConsistencyLevel;
+        private final ConsistencyLevel readConsistencyLevel;
+
+        private AvailabiltyTester(ConsistencyLevel writeConsistencyLevel, ConsistencyLevel readConsistencyLevel)
+        {
+            this.writeConsistencyLevel = writeConsistencyLevel;
+            this.readConsistencyLevel = readConsistencyLevel;
+        }
+
+        public void test(ICoordinator coordinator, int numNodesDown)
+        {
+            if (numNodesDown > maxNodesDown(writeConsistencyLevel) ||
+                numNodesDown > maxNodesDown(readConsistencyLevel))
+                return;
+
+            UUID partitionKey = UUID.randomUUID();
```

Comment:
@adelapena We're testing with a single key here, given RF=3 anyway?

Reply:
One single partition key per test case run, and each `AvailabiltyTester` can be run multiple times. `ONE-ALL` and `ALL-ONE` are run twice, and `QUORUM-QUORUM` is run four times.